File: ios/JetStream/Services/FlightService.swift
import Foundation

struct FlightResponse: Codable, Identifiable {
    let id: String?
    let flightNumber: String
    let airlineCode: String
    let airlineName: String
    let departureAirport: String
    let departureAirportName: String?
    let departureCity: String?
    let arrivalAirport: String
    let arrivalAirportName: String?
    let arrivalCity: String?
    let scheduledDeparture: Date
    let scheduledArrival: Date
    let actualDeparture: Date?
    let actualArrival: Date?
    let status: String?
    let aircraftType: String?
    let seatNumber: String?
    let cabinClass: String?
    let bookingReference: String?
    let departureTerminal: String?
    let arrivalTerminal: String?
    let departureGate: String?
    let arrivalGate: String?
    let distanceKm: Double?
    let durationMinutes: Int?
    let notes: String?
    let isManualEntry: Bool?
    let pnr: String?
    let createdAt: Date?
    let updatedAt: Date?

    // Display helpers
    var route: String {
        "\(departureAirport) → \(arrivalAirport)"
    }

    var durationFormatted: String {
        guard let mins = durationMinutes, mins > 0 else { return "--" }
        return "\(mins / 60)h \(mins % 60)m"
    }

    var isUpcoming: Bool {
        scheduledDeparture > Date()
    }

    var isPast: Bool {
        status == "landed" || scheduledArrival < Date()
    }

    var flightStatus: FlightStatus {
        FlightStatus(rawValue: status ?? "scheduled") ?? .scheduled
    }

    var cabin: CabinClass {
        CabinClass(rawValue: cabinClass ?? "economy") ?? .economy
    }

    var displayId: String {
        id ?? UUID().uuidString
    }
}

struct FlightStatusResponse: Codable {
    let status: String
    let actualDeparture: Date?
    let actualArrival: Date?
    let departureGate: String?
    let arrivalGate: String?
    let departureTerminal: String?
    let arrivalTerminal: String?
    let delayMinutes: Int?
}

// MARK: - Analytics Response Models

struct AnalyticsSummary: Codable {
    let totalFlights: String
    let totalDistanceKm: String
    let totalMinutes: String
    let uniqueAirports: String
    let uniqueAirlines: String

    var flightCount: Int { Int(totalFlights) ?? 0 }
    var distanceKm: Double { Double(totalDistanceKm) ?? 0 }
    var totalHours: Int { (Int(totalMinutes) ?? 0) / 60 }
}

struct AirlineBreakdown: Codable, Identifiable {
    let airlineName: String
    let airlineCode: String
    let flightCount: String
    let totalDistance: String

    var id: String { airlineCode }
    var count: Int { Int(flightCount) ?? 0 }
}

struct AirportStat: Codable, Identifiable {
    let airport: String
    let city: String?
    let visitCount: String

    var id: String { airport }
    var count: Int { Int(visitCount) ?? 0 }
}

struct TrendData: Codable, Identifiable {
    let month: String
    let flightCount: String

    var id: String { month }
    var count: Int { Int(flightCount) ?? 0 }
}

struct RecordsData: Codable {
    let longestFlight: RecordFlight?
    let shortestFlight: RecordFlight?
    let mostFlownRoute: MostFlownRoute?
}

struct RecordFlight: Codable {
    let flightNumber: String
    let airlineName: String
    let departureAirport: String
    let arrivalAirport: String
    let distanceKm: String?

    var route: String { "\(departureAirport) → \(arrivalAirport)" }
}

struct MostFlownRoute: Codable {
    let route: String
    let count: String

    var flightCount: Int { Int(count) ?? 0 }
}

// MARK: - Profile Response

struct ProfileResponse: Codable {
    let id: String
    let email: String
    let name: String
    let avatarUrl: String?
    let authProvider: String?
    let createdAt: Date?
}

final class FlightService {
    private let apiClient = APIClient()

    // MARK: - Flight CRUD

    func getFlights(page: Int = 1, limit: Int = 50) async throws -> [FlightResponse] {
        try await apiClient.request(FlightEndpoint.list(page: page, limit: limit))
    }

    func createFlight(_ request: FlightCreateRequest) async throws -> FlightResponse {
        try await apiClient.request(FlightEndpoint.create(request))
    }

    func deleteFlight(id: String) async throws {
        try await apiClient.request(FlightEndpoint.delete(id: id))
    }

    func getFlightStatus(id: String) async throws -> FlightStatusResponse {
        try await apiClient.request(FlightEndpoint.status(id: id))
    }

    // MARK: - Search

    func searchByFlightNumber(_ flightNumber: String, date: String? = nil) async throws -> [FlightResponse] {
        try await apiClient.request(FlightEndpoint.searchByNumber(flightNumber: flightNumber, date: date))
    }

    func searchAirports(query: String) async throws -> [Airport] {
        try await apiClient.request(FlightEndpoint.searchAirports(query: query))
    }

    func searchAirlines(query: String) async throws -> [Airline] {
        try await apiClient.request(FlightEndpoint.searchAirlines(query: query))
    }

    // MARK: - Analytics

    func getAnalyticsSummary() async throws -> AnalyticsSummary {
        try await apiClient.request(AnalyticsEndpoint.summary)
    }

    func getAirlineBreakdown() async throws -> [AirlineBreakdown] {
        try await apiClient.request(AnalyticsEndpoint.airlines)
    }

    func getAirportStats() async throws -> [AirportStat] {
        try await apiClient.request(AnalyticsEndpoint.airports)
    }

    func getTrends() async throws -> [TrendData] {
        try await apiClient.request(AnalyticsEndpoint.trends)
    }

    func getRecords() async throws -> RecordsData {
        try await apiClient.request(AnalyticsEndpoint.records)
    }

    // MARK: - Profile

    func getProfile() async throws -> ProfileResponse {
        try await apiClient.request(UserEndpoint.getProfile)
    }
}

File: backend/services/flight-service/src/services/flightService.ts
import { query } from '../../../../shared/database/connection';
import { AmadeusService } from './amadeusService';
import { AviationStackService } from './aviationStackService';

const amadeusService = new AmadeusService();
const aviationStackService = new AviationStackService();

export class FlightService {
  async lookupPNR(pnr: string) {
    // Try Amadeus first for PNR details
    try {
      return await amadeusService.lookupPNR(pnr);
    } catch {
      throw new Error(`Could not find flight for PNR: ${pnr}`);
    }
  }

  async getUserFlights(userId: string, page: number, limit: number) {
    const offset = (page - 1) * limit;
    const [flightsResult, countResult] = await Promise.all([
      query(
        'SELECT * FROM flights WHERE user_id = $1 ORDER BY scheduled_departure DESC LIMIT $2 OFFSET $3',
        [userId, limit, offset]
      ),
      query('SELECT COUNT(*) FROM flights WHERE user_id = $1', [userId]),
    ]);

    return {
      flights: flightsResult.rows,
      total: parseInt(countResult.rows[0].count),
    };
  }

  async getFlightById(id: string, userId: string) {
    const result = await query(
      'SELECT * FROM flights WHERE id = $1 AND user_id = $2',
      [id, userId]
    );
    return result.rows[0] || null;
  }

  async createFlight(userId: string, data: any) {
    const result = await query(
      `INSERT INTO flights (user_id, pnr, flight_number, airline_code, airline_name,
        departure_airport, departure_city, arrival_airport, arrival_city,
        scheduled_departure, scheduled_arrival, actual_departure, actual_arrival,
        status, aircraft_type, seat_number, cabin_class, booking_reference,
        distance_km, duration_minutes, departure_terminal, arrival_terminal,
        departure_gate, arrival_gate, notes, is_manual_entry)
       VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14, $15, $16, $17, $18, $19, $20, $21, $22, $23, $24, $25, $26)
       RETURNING *`,
      [
        userId, data.pnr, data.flight_number, data.airline_code, data.airline_name,
        data.departure_airport, data.departure_city, data.arrival_airport, data.arrival_city,
        data.scheduled_departure, data.scheduled_arrival, data.actual_departure || null, data.actual_arrival || null,
        data.status || 'scheduled', data.aircraft_type || null, data.seat_number || null,
        data.cabin_class || 'economy', data.booking_reference || null,
        data.distance_km || null, data.duration_minutes || null,
        data.departure_terminal || null, data.arrival_terminal || null,
        data.departure_gate || null, data.arrival_gate || null,
        data.notes || null, data.is_manual_entry || false,
      ]
    );
    return result.rows[0];
  }

  async updateFlight(id: string, userId: string, data: any) {
    const fields: string[] = [];
    const values: any[] = [];
    let paramIndex = 1;

    const allowedFields = [
      'seat_number', 'cabin_class', 'notes', 'status',
      'actual_departure', 'actual_arrival', 'departure_gate', 'arrival_gate',
      'departure_terminal', 'arrival_terminal', 'booking_reference',
    ];

    for (const field of allowedFields) {
      if (data[field] !== undefined) {
        fields.push(`${field} = $${paramIndex++}`);
        values.push(data[field]);
      }
    }

    if (fields.length === 0) {
      return this.getFlightById(id, userId);
    }

    values.push(id, userId);
    const result = await query(
      `UPDATE flights SET ${fields.join(', ')} WHERE id = $${paramIndex++} AND user_id = $${paramIndex} RETURNING *`,
      values
    );
    return result.rows[0];
  }

  async deleteFlight(id: string, userId: string) {
    await query('DELETE FROM flights WHERE id = $1 AND user_id = $2', [id, userId]);
  }

  async getFlightStatus(id: string) {
    const result = await query('SELECT flight_number, airline_code FROM flights WHERE id = $1', [id]);
    if (!result.rows[0]) throw new Error('Flight not found');

    const { flight_number, airline_code } = result.rows[0];
    try {
      return await aviationStackService.getFlightStatus(flight_number);
    } catch {
      return { status: 'unknown', message: 'Could not fetch live status' };
    }
  }

  async searchAirports(q: string) {
    const result = await query(
      `SELECT * FROM airports WHERE
       iata_code ILIKE $1 OR name ILIKE $2 OR city ILIKE $2
       ORDER BY city LIMIT 10`,
      [q, `%${q}%`]
    );
    return result.rows;
  }

  async searchAirlines(q: string) {
    const result = await query(
      `SELECT * FROM airlines WHERE
       iata_code ILIKE $1 OR name ILIKE $2
       ORDER BY name LIMIT 10`,
      [q, `%${q}%`]
    );
    return result.rows;
  }
}
